### PR TITLE
[Merged by Bors] - fetch: check if hash peer is connected before using it

### LINKED ISSUE #4938: fetcher does not ensure peer is connected before requesting
## Description

code https://github.com/spacemeshos/go-spacemesh/blob/develop/fetch/fetch.go#L486-L490

```go
for _, req := range requests {
		p, exists := f.hashToPeers.GetRandom(req.Hash, req.Hint, rng)
		if !exists {
			p = randomPeer(peers)
		}
                ...
}
```

it gets random peers that registered as hash providers, but it doesn't check if we are connected with that peer.
so for example on my computer i loop with for almost 20 minutes.

2023-08-31T08:41:17.708+0200    WARN    fd68b.fetcher   failed to fetch blocks  {"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "fetcher", "error": "hint: blocksDB, hash: 0xa7fda17e3b073d66aa94e251207faca321e39e7f000000000000000000000000, err: batched request failed w retries: peer is not connected: 12D3KooWCdy5ELnzUoA8kDzG26444DqjSjbNgGMKT3fHPJo7oLmP"}
...
2023-08-31T08:58:01.793+0200	WARN	fd68b.fetcher	failed to fetch blocks	{"node_id": "fd68b9397572556c2f329f3e5af2faf23aef85dbbbb7e38447fae2f4ef38899f", "module": "fetcher", "error": "hint: blocksDB, hash: 0xa7fda17e3b073d66aa94e251207faca321e39e7f000000000000000000000000, err: batched request failed w retries: peer is not connected: 12D3KooWCdy5ELnzUoA8kDzG26444DqjSjbNgGMKT3fHPJo7oLmP"}

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation
Closes #4938

### CODE REVIEW

File: fetch/fetch.go [hunk 1]
```diff
@@ -482,18 +482,28 @@ func (f *Fetch) organizeRequests(requests []RequestMessage) map[p2p.Peer][][]Req
 		}
 		return nil
 	}
+	pm := map[p2p.Peer]struct{}{}
+	for _, p := range peers {
+		pm[p] = struct{}{}
+	}
```
Comment (dshulyak) [inline fetch/fetch.go]: i think it is better to ask host.Connected or something like that. creating a map from peer list on every request is not ideal

**Review by dshulyak — APPROVED** [review]

File: fetch/fetch.go (hunk 1 above)
Comment (countvonzero) [reply fetch/fetch.go]: thanks. will add a method that checks fh.Network().Connectedness(p) == network.Connected

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors cancel

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

File: fetch/cache.go [hunk 2]
```diff
@@ -71,24 +72,15 @@ func (hpc *HashPeersCache) Add(hash types.Hash32, peer p2p.Peer) {
...
-	for peer := range hashPeersMap {
-		i++
-		if i == n {
-			return peer, true
-		}
-	}
-	return p2p.NoPeer, false
+	pm, _ := hpc.getWithStats(hash, hint)
+	peers := maps.Keys(pm)
+	rng.Shuffle(len(peers), func(i, j int) { peers[i], peers[j] = peers[j], peers[i] })
```
Comment (fasmat) [inline fetch/cache.go:82]: `maps.Keys` extracts the keys from a `map` by ranging over it. `range` for maps doesn't produce a deterministic order:

https://go.dev/play/p/nImOnOek6c3

So shuffling might not be necessary.

**Pull request closed without merging**